src/main/java/server/maps/PortalFactory.java

```diff
@@ -33,34 +34,48 @@ public PortalFactory() {
         nextDoorPortal = 0x80;
     }
 
-    public Portal makePortal(int type, Data portal) {
+    public Portal makePortal(int type, int mapId, Data portal) {
         GenericPortal ret = null;
         if (type == Portal.MAP_PORTAL) {
             ret = new MapPortal();
         } else {
             ret = new GenericPortal(type);
         }
-        loadPortal(ret, portal);
+        loadPortal(ret, mapId, portal);
         return ret;
     }
 
-    private void loadPortal(GenericPortal myPortal, Data portal) {
+    private void loadPortal(GenericPortal myPortal, int mapId, Data portal) {
         myPortal.setName(DataTool.getString(portal.getChildByPath("pn")));
         myPortal.setTarget(DataTool.getString(portal.getChildByPath("tn")));
         myPortal.setTargetMapId(DataTool.getInt(portal.getChildByPath("tm")));
         int x = DataTool.getInt(portal.getChildByPath("x"));
         int y = DataTool.getInt(portal.getChildByPath("y"));
         myPortal.setPosition(new Point(x, y));
         String script = DataTool.getString("script", portal, null);
-        if (script != null && script.equals("")) {
-            script = null;
-        }
-        myPortal.setScriptName(script);
         if (myPortal.getType() == Portal.DOOR_PORTAL) {
             myPortal.setId(nextDoorPortal);
             nextDoorPortal++;
         } else {
             myPortal.setId(Integer.parseInt(portal.getName()));
         }
+
+        // If no script node is set, try to find the fallback script and set it
+        // otherwise, don't set the script name. Portals break if the script is not found
+        // and a portal name is set.
+        if (script != null && (!script.isBlank() && !script.isEmpty())) {
+            myPortal.setScriptName(script);
+            return;
+        }
+
+        // Check for and set fallback portal script name if available
+        String fallbackScriptPath = String.format("%s.%s", mapId, portal.getName());
```

Comment:
A period is not ideal as a separator, I would rather see a dash symbol, or maybe an underscore.